### PR TITLE
Change the match-clause cloner

### LINKED ISSUE #26446: Compiler crashes inside match statement
**Description:**
 $title

ballerina-internal.log

`
2020-10-21 08:09:19,774] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:2403)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation$BLangActionInvocation.accept(BLangInvocation.java:218)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExpr(CodeAnalyzer.java:3080)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:1708)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:53)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:1820)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:46)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.lambda$visit$8(CodeAnalyzer.java:723)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:722)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:58)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:826)
	at org.wso2.ballerinalang.compiler.tree.clauses.BLangMatchClause.accept(BLangMatchClause.java:49)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:802)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangMatchStatement.accept(BLangMatchStatement.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.lambda$visit$8(CodeAnalyzer.java:723)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:722)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:58)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:1480)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangForeach.accept(BLangForeach.java:105)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:507)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:58)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visitFunction(CodeAnalyzer.java:477)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:444)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:353)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.lambda$analyzeTopLevelNodes$1(CodeAnalyzer.java:342)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeTopLevelNodes(CodeAnalyzer.java:342)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:337)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:158)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyze(CodeAnalyzer.java:326)
	at org.wso2.ballerinalang.compiler.CompilerDriver.codeAnalyze(CompilerDriver.java:337)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:263)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:250)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:142)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:176)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:220)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:101)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:109)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:65)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.BuildCommand.execute(BuildCommand.java:436)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:57)
`

**Steps to reproduce:**

`import ballerina/http;
import ballerina/io;

http:Client clientEndpoint = new ("http://postman-echo.com");

public function main() {

    string[] animals = ["Mouse"];

    foreach string animal in animals {

        match animal {
            "Mouse" => {
                io:println("Mouse");             
                var response = clientEndpoint->post("/post", "POST: Hello World");
            }
            _ => {
                io:println("Match All");
            }
        }
    }
}
`

**Affected Versions:**
slp4

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Fixes #26446 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java [hunk 1]
```diff
@@ -695,7 +695,7 @@ public void visit(BLangMatchClause source) {
         source.cloneRef = clone;
         clone.matchPatterns = cloneList(source.matchPatterns);
         clone.setMatchGuard(source.getMatchGuard());
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java:697]: shouldn't we clone here as well?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal [hunk 2]
```diff
@@ -405,6 +405,23 @@ function testConstPattern12() {
     assertEquals("OTHER", constPattern12(3));
 }
 
+client class Client {
+    remote function post(string x, string y) returns anydata {
+    }
+}
+
+Client clientEndpoint = new ();
+public function testConstPattern13() returns string {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:414]: Ideally, we should validate something here, right? Shall have `post` return some value, and assert it inside the match, and remove `returns string` here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java (hunk 1 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java]: yes, fixed it

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal (hunk 2 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal]: changed.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal [hunk 3]
```diff
@@ -405,6 +405,23 @@ function testConstPattern12() {
...
+}
+
+Client clientEndpoint = new ();
+public function testConstPattern13() {
+    string animal = "Mouse";
+    match animal {
+        "Mouse" => {
+            var res = clientEndpoint->post("/post", "POST: Hello World");
+            assertEquals("POST: Hello World", res);
+        }
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:421]: We should probably have a wildcard match pattern here and panic if it reaches that block, just for completeness.

**Review by MaryamZi — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal (hunk 3 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/const-pattern.bal:421]: Fixed.

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**